### ISSUE TITLE
machined: access denied, copy-from works anyway

### ISSUE THREAD

**Issue by jernst** [systemd/systemd#433]
```
$ sudo machinectl copy-from webapptest-20150630-162953 /home/shepherd/dot.dot $(pwd)/dot.dot
Failed to copy: Access denied
# exit code 1

$ ls -al dot.dot
-rw-r--r-- 1 root root 32818 Jun 30 17:35 dot.dot
```

Host: Arch, systemd 221-2; container: UBOS (like Arch), systemd 219.

**Comment by poettering** [comment]
Any chance you can "strace -s 500 -o /tmp/log -p $PID" the machined process while you execute this? The EPERM issue must come from somewhere, the strace might be able to show us that.

**Comment by jernst** [comment]
Here's the `/tmp/log` as requested:

```
epoll_wait(4, {{EPOLLIN, {u32=3330637088, u64=140663509581088}}}, 3, -1) = 1
clock_gettime(CLOCK_BOOTTIME, {152747, 752050458}) = 0
socket(PF_LOCAL, SOCK_DGRAM|SOCK_CLOEXEC, 0) = 8
sendmsg(8, {msg_name(21)={sa_family=AF_LOCAL, sun_path="/run/systemd/notify"}, msg_iov(1)=[{"WATCHDOG=1", 10}], msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 10
close(8)                                = 0
timerfd_settime(5, TFD_TIMER_ABSTIME, {it_interval={0, 0}, it_value={152792, 591423000}}, NULL) = 0
recvmsg(6, {msg_name(0)=NULL, msg_iov(1)=[{"l\1\4\1A\0\0\0\2\0\0\0\266\0\0\0\1\1o\0\31\0\0\0", 24}], msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_NOSIGNAL|MSG_CMSG_CLOEXEC) = 24
recvmsg(6, {msg_name(0)=NULL, msg_iov(1)=[{"/org/freedesktop/machine1\0\0\0\0\0\0\0\3\1s\0\17\0\0\0CopyFromMachine\0\2\1s\0 \0\0\0org.freedesktop.machine1.Manager\0\0\0\0\0\0\0\0\6\1s\0\30\0\0\0org.freedesktop.machine1\0\0\0\0\0\0\0\0\10\1g\0\3sss\0\0\0\0\0\0\0\0\7\1s\0\5\0\0\0:1.64\0\0\0\5\0\0\0depot\0\0\0#\0\0\0/home/shepherd/.ssh/authorized_keys\0\10\0\0\0/tmp/foo\0", 241}], msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_NOSIGNAL|MSG_CMSG_CLOEXEC) = 241
sendmsg(6, {msg_name(0)=NULL, msg_iov(2)=[{"l\1\0\1\n\0\0\0\35\0\0\0\207\0\0\0\1\1o\0\25\0\0\0/org/freedesktop/DBus\0\0\0\3\1s\0\25\0\0\0GetConnectionUnixUser\0\0\0\2\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\6\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\10\1g\0\1s\0\0", 152}, {"\5\0\0\0:1.64\0", 10}], msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 162
recvmsg(6, {msg_name(0)=NULL, msg_iov(1)=[{"l\2\1\1\4\0\0\0\21\0\0\0=\0\0\0\6\1s\0\5\0\0\0", 24}], msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_NOSIGNAL|MSG_CMSG_CLOEXEC) = 24
recvmsg(6, {msg_name(0)=NULL, msg_iov(1)=[{":1.47\0\0\0\5\1u\0\35\0\0\0\10\1g\0\1u\0\0\7\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\0\0\0\0", 60}], msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_NOSIGNAL|MSG_CMSG_CLOEXEC) = 60
getuid()                                = 0
open("/tmp", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_CLOEXEC) = 8
pipe2([9, 11], O_NONBLOCK|O_CLOEXEC)    = 0
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7feec5dfda50) = 21077
close(11)                               = 0
signalfd4(10, [CHLD], 8, O_NONBLOCK|O_CLOEXEC) = 10
close(8)                                = 0
epoll_wait(4, {{EPOLLIN, {u32=6, u64=6}}}, 4, 0) = 1
clock_gettime(CLOCK_BOOTTIME, {152747, 767134799}) = 0
read(10, "\21\0\0\0\0\0\0\0\1\0\0\0UR\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 128) = 128
waitid(P_PID, 21077, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=21077, si_uid=0, si_status=1, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
read(10, 0x7ffddd729fb0, 128)           = -1 EAGAIN (Resource temporarily unavailable)
signalfd4(10, [], 8, O_NONBLOCK|O_CLOEXEC) = 10
read(9, "\377\377\377\377", 4)          = 4
sendmsg(6, {msg_name(0)=NULL, msg_iov(2)=[{"l\3\1\1\34\0\0\0\36\0\0\0O\0\0\0\5\1u\0\2\0\0\0\6\1s\0\5\0\0\0:1.64\0\0\0\4\1s\0'\0\0\0org.freedesktop.DBus.Error.AccessDenied\0\10\1g\0\1s\0\0", 96}, {"\27\0\0\0Operation not permitted\0", 28}], msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 124
close(9)                                = 0
waitid(P_PID, 21077, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=21077, si_uid=0, si_status=1, si_utime=0, si_stime=0}, WNOHANG|WEXITED, NULL) = 0
epoll_wait(4,  <detached ...>
```

**Comment by jernst** [comment]
This may be related to file ownership in the container: the file in the original report is owned by user shepherd, not root. Here's a test:

In the container:

```
# ls -al A B C D
-rw-r--r-- 1 root     root 2 Jul 12 19:12 A
---------- 1 root     root 2 Jul 12 19:12 B
-rw-r--r-- 1 shepherd root 2 Jul 12 19:12 C
---------- 1 shepherd root 2 Jul 12 19:13 D
```

On the host:

```
# machinectl copy-from depot /tmp/A /tmp/A
# machinectl copy-from depot /tmp/B /tmp/B
# machinectl copy-from depot /tmp/C /tmp/C
Failed to copy: Access denied
# machinectl copy-from depot /tmp/D /tmp/D
# ls -al A B C D
-rw-r--r-- 1 root root 2 Jul 12 19:12 A
---------- 1 root root 2 Jul 12 19:12 B
-rw-r--r-- 1 root root 2 Jul 12 19:12 C
---------- 1 root root 2 Jul 12 19:13 D
```

I tried a few other things, and it seems unrelated to the ownership on the containing directory; only the ownership of the to-be-copied file seems to matter.

**Comment by poettering** [comment]
Hmm, could you redo the strace, but this time add "-f" to the strace cmdline flags?

Do you use user namespaces (i.e. nspawn's --private-users switch)?

**Comment by poettering** [comment]
Oh, yuck, I think I know what is going on...

can you change your systemd-machined.service to include CAP_CHOWN, CAP_FOWNER, CAP_FSETID int the list of caps in the CapabilityBoundingSet= line? THen, restart machined. Does it work then? If so, I guess we just need to add these three caps to the service file and should be good.

Please verify and report back!

**Comment by jernst** [comment]
Correct! Adding those capabilities makes the error message go away, and the new file has the correct owner and group.

**Comment by poettering** [comment]
There's a fix now in PR #738.
